### PR TITLE
Fix FAISS SQ merges with vectorless input segments

### LINKED ISSUE #3379: [BUG] FAISS SQ merge can fail when empty scalar-quantized vector values lack quantizedVectorValues
**What is the bug?**
For native FAISS scalar-quantized vector fields, segment merge can fail when Lucene returns an empty vector-values implementation for a field.

The affected path reopens merged flat vector files and calls `KNN1040ScalarQuantizedUtils.extractQuantizedByteVectorValues(...)`. That utility reflects a private Lucene field named `quantizedVectorValues` from the concrete runtime class. When a merge input segment has no live vectors for the field, Lucene may return an empty/off-heap wrapper that does not declare that field. This can throw `NoSuchFieldException`, fail `QuantizedByteVectorValues` extraction, and abort the merge.

Observed failure signature:
```text
merge_exception: this writer hit an unrecoverable error; cannot merge
Caused by: IOException: Failed to extract QuantizedByteVectorValues from floatVectorValues [EmptyOffHeapVectorValues/...]
Caused by: NoSuchFieldException: quantizedVectorValues
```

**How can one reproduce the bug?**

A minimal reproduction should create a FAISS scalar-quantized knn_vector field, produce multiple segments, ensure at least one merge input segment has zero live values for that field, then trigger a merge.

Example shape:
1. Create an index with a FAISS HNSW knn_vector field using an SQ encoder.
2. Index documents across multiple segments, including documents that omit the vector field or later delete all documents containing that field in one segment.
3. Trigger merge or force-merge.
4. The merge path may receive empty FloatVectorValues for the SQ field and fail while extracting quantizedVectorValues.

**What is the expected behavior?**
If a merge input has no live vectors for a scalar-quantized field, k-NN should treat it as empty and skip quantized extraction/native build work for that field instead of failing the merge.

For non-empty values, quantized extraction should continue to work. If reflection is still needed, lookup should be robust to Lucene wrappers whose private field is declared in a superclass.

**What is your host/environment?**
- OpenSearch: 3.6.x
- k-NN plugin: 3.6.x
- Engine: FAISS
- Vector field type: knn_vector
- Method: HNSW
- Encoder: SQ / scalar quantization

**Do you have any screenshots?**
No screenshots.

**Do you have any additional context?**

Relevant code paths:

- `Faiss1040ScalarQuantizedKnnVectorsWriter.flush`
- `Faiss1040ScalarQuantizedKnnVectorsWriter.mergeOneField`
- `KNN1040ScalarQuantizedUtils.extractQuantizedByteVectorValues`

A defensive fix would:
- short-circuit when FloatVectorValues is null or size() == 0,
- search the class hierarchy for quantizedVectorValues instead of only the concrete runtime class,
- add test coverage for empty values and superclass field lookup.

Example sketch: 
``` java

final FloatVectorValues floatVectorValues = flatVectorsReader.getFloatVectorValues(fieldInfo.getName());
if (floatVectorValues == null || floatVectorValues.size() == 0) {
    log.debug("No scalar-quantized vectors found for field {}", fieldInfo.getName());
    return;
}

final QuantizedByteVectorValues quantizedValues =
    KNN1040ScalarQuantizedUtils.extractQuantizedByteVectorValues(floatVectorValues);
```

``` java
private static Field findFieldInClassHierarchy(final Class<?> candidateClass, final String fieldName) {
    Class<?> currentClass = candidateClass;
    while (currentClass != null) {
        try {
            return currentClass.getDeclaredField(fieldName);
        } catch (NoSuchFieldException ignored) {
            currentClass = currentClass.getSuperclass();
        }
    }
    return null;
}
```

### PR DESCRIPTION
**Pull request by JeremyRitchie**
### Description

FAISS scalar-quantized segment merges can fail when a segment that declares the vector field contains documents but no vector values. Lucene returns `EmptyOffHeapVectorValues` for that source segment; the type does not contain the private `quantizedVectorValues` field expected by `Faiss1040ScalarQuantizedFlatVectorsReader`, so reflective extraction throws before the merged-output guard added in #3381 can run.

This change preserves the reader's existing `ScalarQuantizedFloatVectorValues` contract by supplying an empty `QuantizedByteVectorValues` implementation whenever the delegate reports zero values, before attempting reflective extraction. Non-empty values retain the original extraction path, so incompatibilities still propagate instead of silently omitting the native index. Invalid vector access on the empty implementation follows Lucene's empty off-heap behavior by throwing `UnsupportedOperationException`.

The codec regression test writes a real one-document segment with no vector value, verifies its wrapped empty read path and warmup behavior, merges it with a segment containing 150 vectors, and verifies every surviving vector remains readable. An integration test also force-merges ten non-vector documents to one segment, adds ten vector documents in a second segment, force-merges again, and verifies k-NN search.

This is a Lucene 10.5 adaptation of the system-tested fix developed with @alanshieh.

### Related Issues

Resolves #3379

Follow-up to #3381.

### Check List

- [x] New functionality includes testing.
- [x] New functionality has been documented (code-level behavior; no public API change).
- [x] API changes companion pull request (not applicable; no API change).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR (not applicable; internal merge correctness fix).

### Testing

- [x] `./gradlew spotlessApply spotlessJavaCheck`
- [x] Focused wrapper, reader, warmup, and writer tests: 44 tests, 0 failures
- [x] Requested force-merge integration regression: 1 test, 0 failures (seed `4F233982CC831F4D`)
- [x] `./gradlew precommit`
- [x] Initial patch: `./gradlew build -Dnproc.count=4` (`BUILD SUCCESSFUL` in 2h 10m 59s; seed `8C02CBE5BB25AF28`)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedFlatVectorsReader.java [hunk 1]
```diff
@@ -67,14 +70,32 @@ public void checkIntegrity() throws IOException {
...
-        );
+        if (floatVectorValues == null) {
+            return null;
+        }
+
+        final QuantizedByteVectorValues quantizedValues;
+        try {
+            quantizedValues = KNN1040ScalarQuantizedUtils.extractQuantizedByteVectorValues(floatVectorValues);
+        } catch (IOException e) {
+            if (e.getCause() instanceof NoSuchFieldException && floatVectorValues.size() == 0) {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedFlatVectorsReader.java]: rather than getting and exception can we just add a top level if condition if `floatVectorValues.size() == 0` then just return `return new ScalarQuantizedFloatVectorValues(floatVectorValues, new EmptyQuantizedByteVectorValues(floatVectorValues));` and rest of the flow which was there originally can stay?

**Review by navneet1v — COMMENTED** [review]
Can we also add an IT like this:
1. create a vector index 
2. Ingest 10 docs without vectors in the index.
3. Force merge to 1 segment.
4. Then index 10 more new docs with vectors. which will create a new segment.
5. Do force merge. and then do search.

This should help validate the bug and also validate the fix is actually helping.

File: src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public int size() {
+        return floatVectorValues.size();
+    }
+
+    @Override
+    public byte[] vectorValue(int ord) throws IOException {
+        return emptyVector;
+    }
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java]: is this the same contract which lucene has for EmptyFloatVectorValues?

File: src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedFlatVectorsReader.java (hunk 1 above)
Comment (JeremyRitchie) [reply src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedFlatVectorsReader.java]: Done

File: src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java (hunk 2 above)
Comment (JeremyRitchie) [reply src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java]: Good catch.
I checked Lucene 10.5, and our behavior wasn’t quite the same. Lucene’s empty off-heap vector values throw UnsupportedOperationException if vectorValue() is called, while ours returned an empty byte array. I’ve updated our implementation to match Lucene and adjusted the unit test accordingly.

File: src/test/java/org/opensearch/knn/integ/ModeAndCompressionIT.java [hunk 3]
```diff
@@ -583,6 +583,56 @@ public void testMixedSegmentsWithNonVectorDoc_whenValid_ThenSucceed() {
         }
     }
 
+    /**
+     * Regression test for merging a vectorless segment with a FAISS 1-bit scalar-quantized segment.
+     * First force-merges non-vector documents into one segment, then adds vector documents in a
+     * second segment and force-merges both segments before searching.
+     */
+    @SneakyThrows
+    public void testForceMergeWithVectorlessSegment_whenFaissSQ_thenSucceed() {
```
Comment (navneet1v) [inline src/test/java/org/opensearch/knn/integ/ModeAndCompressionIT.java]: curious to know without your fix this IT is able to capture the issue or not?

File: src/test/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedKnnVectorsWriterTests.java [hunk 4]
```diff
@@ -434,8 +437,131 @@ public void testMergeOneField_whenFlatReaderReturnsEmptyVectors_thenSkipsNativeB
...
+    public void testMergeOneField_whenInputSegmentHasNoVectors_thenSucceeds() {
+        final int vectorsPerSegment = 150;
+        final int vectorlessSegmentDocs = 1;
+        final float[][] vectors = generateRandomVectors(vectorsPerSegment, DIMENSION);
+        final FieldInfo fi = createRealFieldInfo();
+        final FieldInfos fieldInfos = new FieldInfos(new FieldInfo[] { fi });
+        final Faiss1040ScalarQuantizedKnnVectorsFormat format = new Faiss1040ScalarQuantizedKnnVectorsFormat();
+        final Lucene104ScalarQuantizedVectorsFormat luceneSqFormat = new Lucene104ScalarQuantizedVectorsFormat(SINGLE_BIT_QUERY_NIBBLE);
+
+        try (org.apache.lucene.store.Directory directory = newDirectory()) {
```
Comment (navneet1v) [inline src/test/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedKnnVectorsWriterTests.java]: [nit-pick] can we just export this class?

**Review by navneet1v — DISMISSED** [review]
Overall looks good. Minor questions

**Comment by navneet1v** [comment]
added skip change log label since changelog verifier was failing even when changelog entry is present

**Comment by navneet1v** [comment]
@JeremyRitchie why did we remove the IT?

File: src/test/java/org/opensearch/knn/integ/ModeAndCompressionIT.java (hunk 3 above)
Comment (JeremyRitchie) [reply src/test/java/org/opensearch/knn/integ/ModeAndCompressionIT.java]: I checked this by temporarily removing the size() == 0 reader guard and running the IT. It still passed, so it wasn’t actually reproducing this failure. A segment containing only documents that never had vectors doesn’t create the empty SQ vector metadata that causes Lucene to return EmptyOffHeapVectorValues. I removed the IT rather than leave misleading regression coverage. 

The real-I/O codec test `testMergeOneField_whenInputSegmentHasNoVectors_thenSucceeds` **does** reproduce the issue: without the guard it fails with an IOException caused by NoSuchFieldException: quantizedVectorValues, and with the fix it passes.

File: src/test/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedKnnVectorsWriterTests.java (hunk 4 above)
Comment (JeremyRitchie) [reply src/test/java/org/opensearch/knn/index/codec/KNN1040Codec/Faiss1040ScalarQuantizedKnnVectorsWriterTests.java]: import? Yup, done.

File: src/test/java/org/opensearch/knn/integ/ModeAndCompressionIT.java (hunk 3 above)
Comment (navneet1v) [reply src/test/java/org/opensearch/knn/integ/ModeAndCompressionIT.java]: @JeremyRitchie but in when you were doing testing in your distribution you were able to repro it, can we not build a similar thing in the IT?

**Comment by JeremyRitchie** [comment]
> @JeremyRitchie why did we remove the IT?


My thinking was that we wanted test coverage to ensure this fix actually works. This was the success criteria as i understood it.

I was able to validate that we have that without the IT, and that the IT was actually ineffective at validating the fix so it seemed rather pointless to me. 

I'm happy to add an IT which will validate this fix - just wanted to keep this lean.

**Comment by navneet1v** [comment]
> > @JeremyRitchie why did we remove the IT?
> 
> My thinking was that we wanted test coverage to ensure this fix actually works. This was the success criteria as i understood it.
> 
> I was able to validate that we have that without the IT, and that the IT was actually ineffective at validating the fix so it seemed rather pointless to me.
> 
> I'm happy to add an IT which will validate this fix - just wanted to keep this lean.

We want a test that can repro the error and the fix shows the failing test worked.

**Comment by navneet1v** [comment]
> > > @JeremyRitchie why did we remove the IT?
> > 
> > 
> > My thinking was that we wanted test coverage to ensure this fix actually works. This was the success criteria as i understood it.
> > I was able to validate that we have that without the IT, and that the IT was actually ineffective at validating the fix so it seemed rather pointless to me.
> > I'm happy to add an IT which will validate this fix - just wanted to keep this lean.
> 
> We want a test that can repro the error and the fix shows the failing test worked.

Yes lets have it as an IT also. IT generally survives more than UTs in the repo as per experience working in KNN repo. But its not a blocker for merge

**Comment by JeremyRitchie** [comment]
> Yes lets have it as an IT also. IT generally survives more than UTs in the repo as per experience working in KNN repo. But its not a blocker for merge

Added the fixed IT.
- Without the reader guard, it reproduces the quantizedVectorValues merge failure
- With the fix restored, the IT passes

File: src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * An empty implementation of {@link QuantizedByteVectorValues} used when Lucene returns empty
+ * vector values that do not expose the {@code quantizedVectorValues} field.
+ *
+ * <p>This class delegates size and iteration operations to the underlying {@link FloatVectorValues},
+ * rejects vector access, and returns empty defaults for quantization-specific operations. It maintains the
+ * {@link ScalarQuantizedFloatVectorValues} return type expected by callers without fabricating
+ * quantized data for a segment that contains no vectors.
+ */
+class EmptyQuantizedByteVectorValues extends QuantizedByteVectorValues {
```
Comment (kotwanikunal) [inline src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java]: Why not use the existing `ScalarQuantizedFloatVectorValues` to do null checks instead? 

https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/ScalarQuantizedFloatVectorValues.java#L54-L56
 and https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/ScalarQuantizedFloatVectorValues.java#L64-L66 can have null guards instead. 

We can reuse the existing class structure and just return null with 0 floatVectorValues


Within the reader - 

```
if (floatVectorValues == null) {
    return null;
}
if (floatVectorValues.size() == 0) {
    return new ScalarQuantizedFloatVectorValues(floatVectorValues, null);
}
```

Comment (JeremyRitchie) [reply src/main/java/org/opensearch/knn/index/codec/KNN1040Codec/EmptyQuantizedByteVectorValues.java]: Sounds good - i've made that change now. :)

**Review by kotwanikunal — APPROVED** [review]

**Review by navneet1v — APPROVED** [review]

**Comment by navneet1v** [comment]
Will merge the change once CI passes
